parsec/mca/device/device.c

```diff
@@ -873,16 +846,68 @@ static int cpu_weights(parsec_device_module_t* device, int nstreams)
         fp_ipc = 16;
         dp_ipc = 8;
     }
-#endif
-#endif  /* defined(__x86_64__) || defined(__i386__) */
+    else {
+        fp_ipc = 8;
+        dp_ipc = 4;
+    }
+#elif defined(__PPC64__)
+    fp_ipc = 16;
+    dp_ipc = 8;
+#endif  /* defined(__x86_64__) || defined(__i386__) || defined(__PPC64__) */
     free(cpu_flags);
 
+#elif defined(__APPLE__)
+    size_t len = sizeof(cpu_model);
+    size_t val = 0;
+    int rc = sysctlbyname("machdep.cpu.brand_string", cpu_model, &len, NULL, 0);
+    if( rc ) {
+        parsec_warning("CPU Name cannot be autodetected on this machine (OSX): %s", strerror(errno));
+        goto notfound;
+    }
+    /* is it an arm64 mac? */
+    rc = sysctlbyname("hw.optional.arm64", &val, &len, NULL, 0);
+    if( 0 == rc && val ) {
+        /* vector length */
+        fp_ipc = 16;
+        dp_ipc = 8;
+        // TODO: figure out a way to obtain cpu freq on arm64 macos
+        freq = 0.0;
+    }
+    else { /* intel mac */
+        /* vector length */
+        int i;
+        char *keys[4] = {
+          "hw.optional.avx512f",
+          "hw.optional.avx2_0",
+          "hw.optional.avx1_0",
+          NULL
+        };
+        for(i = 0; keys[i] != NULL; i++) {
+            rc = sysctlbyname(keys[i], &val, &len, NULL, 0);
+            if( 0 == rc && val )
+                break;
+        }
+        fp_ipc = 8*(8>>i);
+        dp_ipc = 4*(8>>i);
+        /* frequency */
+        len = sizeof(size_t);
+        rc = sysctlbyname("hw.cpufrequency", &val, &len, NULL, 0);
+        if( rc ) {
+            parsec_warning("CPU Frequency cannot be autodetected on this machine (OSX): %s", strerror(errno));
+            goto notfound;
+        }
+        freq = val * 1e-9f;
+    }
+#endif
+
+notfound:
```

Comment:
I can see that this code is adding support for arm64, but it also replace the cross-OS AVX detection with a single OSX specific detection. How are the different AVX flavors detected for Linux (the entire detection based on __builtin_cpu_supports being removed) ?

Reply:
The builtin_cpu_support always return no support because it requires we build parsec with specific flags that we never pass. 

The linux generic discovery is done by reading proc/cpuinfo above on lines 864/837 (old/new)

Reply:
I'm not sure I understand. `__builtin_cpu_support` returns based on the current processor capabilities, which is what we want. Anyway, the OSX support via `sysctlbyname` is valid and more generic (does not require GNU compilers) so it is a step in the right direction.